File: examples/runtimes/csharp/src/NodeRegistry.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Text.Json;

namespace Blok.Runtime;

/// <summary>
/// NodeRegistry manages registered node handlers and provides execution
/// capabilities with error handling and performance metrics.
/// </summary>
public sealed class NodeRegistry
{
    private readonly ConcurrentDictionary<string, INodeHandler> _nodes = new();

    /// <summary>
    /// Registers a node handler with the given name.
    /// </summary>
    /// <param name="name">The unique name used to identify this node.</param>
    /// <param name="handler">The node handler implementation.</param>
    /// <exception cref="ArgumentNullException">Thrown when name or handler is null.</exception>
    public void Register(string name, INodeHandler handler)
    {
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(handler);
        _nodes[name] = handler;
    }

    /// <summary>
    /// Retrieves a node handler by name.
    /// </summary>
    /// <param name="name">The name of the node to retrieve.</param>
    /// <returns>The node handler, or null if not found.</returns>
    public INodeHandler? Get(string name)
    {
        _nodes.TryGetValue(name, out var handler);
        return handler;
    }

    /// <summary>
    /// Executes a node identified by the execution request.
    /// Captures timing metrics and handles errors gracefully.
    /// </summary>
    /// <param name="request">The execution request from the Blok runner.</param>
    /// <returns>The execution result with data or error details.</returns>
    public async Task<ExecutionResult> ExecuteAsync(ExecutionRequest request)
    {
        var handler = Get(request.Node.Name);

        if (handler is null)
        {
            return ExecutionResult.Fail(
                $"Node '{request.Node.Name}' not found",
                "NodeNotFoundError"
            );
        }

        var stopwatch = Stopwatch.StartNew();

        try
        {
            var data = await handler.ExecuteAsync(request.Context, request.Node.Config);
            stopwatch.Stop();

            var metrics = new ExecutionMetrics
            {
                DurationMs = stopwatch.Elapsed.TotalMilliseconds,
                MemoryBytes = GC.GetTotalMemory(forceFullCollection: false)
            };

            return ExecutionResult.Ok(data, metrics);
        }
        catch (Exception ex)
        {
            stopwatch.Stop();

            return ExecutionResult.Fail(
                ex.Message,
                ex.GetType().Name
            );
        }
    }

    /// <summary>
    /// Returns the health status of the runtime, including the list of
    /// registered node names.
    /// </summary>
    /// <param name="version">The runtime version string.</param>
    /// <returns>A health status object.</returns>
    public HealthStatus GetHealth(string version)
    {
        return new HealthStatus
        {
            Status = "healthy",
            Version = version,
            NodesLoaded = _nodes.Keys.ToList()
        };
    }

    /// <summary>
    /// Returns the number of registered nodes.
    /// </summary>
    public int Count => _nodes.Count;
}

File: examples/runtimes/ruby/lib/node_registry.rb
# frozen_string_literal: true

module Blok
  # NodeRegistry manages registered node handlers.
  #
  # It provides methods to register, retrieve, and execute node handlers,
  # as well as report health status for the runtime.
  class NodeRegistry
    def initialize
      @nodes = {}
    end

    # Register a node handler with the given name.
    #
    # @param name [String] the unique name for this node
    # @param handler [Blok::NodeHandler] the handler instance
    # @return [void]
    def register(name, handler)
      @nodes[name] = handler
    end

    # Retrieve a node handler by name.
    #
    # @param name [String] the node name
    # @return [Blok::NodeHandler] the registered handler
    # @raise [KeyError] if the node is not found
    def get(name)
      unless @nodes.key?(name)
        raise KeyError, "Node '#{name}' not found"
      end

      @nodes[name]
    end

    # Execute a node by processing an ExecutionRequest.
    #
    # Looks up the handler by node name, runs it, and wraps the result
    # in an ExecutionResult. Errors are caught and returned as a failed result.
    #
    # @param request [Blok::ExecutionRequest] the execution request
    # @return [Blok::ExecutionResult] the execution result
    def execute(request)
      handler = get(request.node.name)
      data = handler.execute(request.context, request.node.config)

      ExecutionResult.new(success: true, data: data)
    rescue StandardError => e
      ExecutionResult.new(
        success: false,
        data: nil,
        errors: { "message" => e.message, "type" => e.class.name }
      )
    end

    # Return health status for the runtime.
    #
    # @param version [String] the runtime version string
    # @return [Blok::HealthStatus]
    def health(version)
      HealthStatus.new(
        version: version,
        nodes_loaded: @nodes.keys
      )
    end

    # Return the number of registered nodes.
    #
    # @return [Integer]
    def size
      @nodes.size
    end
  end
end
